### PR TITLE
feat: introduce metrics for failed evaluations

### LINKED ISSUE #559: Metric for evaluations by reason (including errors)
We should create a new metric which buckets _evaluations by reason_, which would include evaluations that failed because they were not found flags, or evaluations which had un-parsable context, or other error conditions, as well as flags that evaluated some way due to targeting vs static assignment, etc.

Created based on [this](https://github.com/open-feature/flagd/pull/556#discussion_r1147789139) (cc @beeme1mr @thisthat )

### PR DESCRIPTION
**Pull request by odubajDT**
## This PR

- introduced a new metric which buckets evaluations by reason
- small refactoring
- unit tests

### Related Issues

Fixes #559 

### How to test

Run flagd via 
```
make run
```
Run
```
curl -X POST "localhost:8013/schema.v1.Service/ResolveString" -d '{"flagKey":"fibAlgo","context":{}}' -H "Content-Type: application/json"
curl -X POST "localhost:8013/schema.v1.Service/ResolveString" -d '{"flagKey":"myStringFlag","context":{}}' -H "Content-Type: application/json"
curl -X POST "localhost:8013/schema.v1.Service/ResolveString" -d '{"flagKey":"myStringFlagNonExisting","context":{}}' -H "Content-Type: application/json"
curl -X POST "localhost:8013/schema.v1.Service/ResolveString" -d '{"flagKey":"myStringFlagNonExisting2","context":{}}' -H "Content-Type: application/json"
curl -X POST "localhost:8013/schema.v1.Service/ResolveInt" -d '{"flagKey":"myStringFlag","context":{}}' -H "Content-Type: application/json"
```

Open http://localhost:8014/metrics and see something like the following

```
# HELP impressions_total The number of evaluation for a given flag
# TYPE impressions_total counter
impressions_total{feature_flag_key="fibAlgo",feature_flag_provider_name="flagd",feature_flag_reason="DEFAULT",feature_flag_variant="recursive",otel_scope_name="openfeature/flagd",otel_scope_version=""} 1
impressions_total{feature_flag_key="myStringFlag",feature_flag_provider_name="flagd",feature_flag_reason="STATIC",feature_flag_variant="key1",otel_scope_name="openfeature/flagd",otel_scope_version=""} 1
```
and

```
# HELP reasons_total The number of evaluations for a given reason
# TYPE reasons_total counter
reasons_total{feature_flag_provider_name="flagd",feature_flag_reason="DEFAULT",otel_scope_name="openfeature/flagd",otel_scope_version=""} 1
reasons_total{feature_flag_provider_name="flagd",feature_flag_reason="STATIC",otel_scope_name="openfeature/flagd",otel_scope_version=""} 1
reasons_total{exception_type="FLAG_NOT_FOUND",feature_flag_provider_name="flagd",feature_flag_reason="ERROR",otel_scope_name="openfeature/flagd",otel_scope_version=""} 2
reasons_total{exception_type="TYPE_MISMATCH",feature_flag_provider_name="flagd",feature_flag_reason="ERROR",otel_scope_name="openfeature/flagd",otel_scope_version=""} 1
```

### CODE REVIEW

**Review by thisthat — CHANGES_REQUESTED** [review]
In the ticket, the agreement was to create a new metric instead of reusing impressions.

File: core/pkg/eval/ievaluator.go [hunk 1]
```diff
@@ -10,14 +12,41 @@ type AnyValue struct {
...
+func (a AnyValue) GetMetricsAttributes() []attribute.KeyValue {
+	if a.Error != nil {
+		return []attribute.KeyValue{
+			semconv.FeatureFlagProviderName("flagd"),
+			{
+				Key:   attribute.Key("reason"),
+				Value: attribute.StringValue(a.Reason),
+			},
+			{
+				Key:   attribute.Key("errType"),
```
Comment (thisthat) [inline core/pkg/eval/ievaluator.go:37]: there is a semconv for errors: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/semantic_conventions/exceptions.md

File: core/pkg/service/flag-evaluation/flag_evaluator.go [hunk 2]
```diff
@@ -283,10 +285,12 @@ func errFormat(err error) error {
 		return connect.NewError(connect.CodeNotFound, fmt.Errorf("%s, %s", ErrorPrefix, err.Error()))
 	case model.TypeMismatchErrorCode:
 		return connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("%s, %s", ErrorPrefix, err.Error()))
-	case model.DisabledReason:
+	case model.FlagDisabledErrorCode:
```
Comment (thisthat) [inline core/pkg/service/flag-evaluation/flag_evaluator.go:288]: why this change?

File: core/pkg/eval/ievaluator.go [hunk 3]
```diff
@@ -2,25 +2,71 @@ package eval
...
+}
+
+func (a AnyValue) GetImpressionAttributes() []attribute.KeyValue {
+	return []attribute.KeyValue{
+		semconv.FeatureFlagProviderName(FlagdProviderName),
+		semconv.FeatureFlagKey(a.FlagKey),
+		semconv.FeatureFlagVariant(a.Variant),
+		FeatureFlagReason(a.Reason),
 	}
 }
```
Comment (thisthat) [inline core/pkg/eval/ievaluator.go:43]: Idk if we should add information about semantic convention directly in the "interface". 
I would rather have semantic conventions in the `metrics.go` or a new file under that package. 
I am thinking about the use case where we could decide to add/remove a dimension from a metric. To implement this, we would need to know which object implements these attributes instead of looking directly nearby where we register the metric datapoint.

File: core/pkg/eval/ievaluator.go [hunk 4]
```diff
@@ -2,25 +2,71 @@ package eval
...
 
+func (a AnyValue) GetReasonAttributes() []attribute.KeyValue {
+	attrs := []attribute.KeyValue{
+		semconv.FeatureFlagProviderName(FlagdProviderName),
+		FeatureFlagReason(a.Reason),
+	}
+
+	if a.HasError() {
+		attrs = append(attrs, ExceptionType(a.Error.Error()))
+	}
```
Comment (thisthat) [inline core/pkg/eval/ievaluator.go:53]: I don't think we need this attribute in the metric. The reason should be smth like one of `STATIC|TARGET_MATCH|ERROR`

File: core/pkg/service/flag-evaluation/flag_evaluator.go (hunk 2 above)
Comment (odubajDT) [reply core/pkg/service/flag-evaluation/flag_evaluator.go:288]: It was a bug, it is a reason and not an error

File: core/pkg/eval/ievaluator.go (hunk 4 above)
Comment (odubajDT) [reply core/pkg/eval/ievaluator.go:53]: yes this is another dimension to give more info about the error, if we do not need it, I will remove it

File: core/pkg/otel/metrics.go [hunk 5]
```diff
@@ -46,12 +53,19 @@ func (r MetricsRecorder) InFlightRequestEnd(ctx context.Context, attrs []attribu
...
-	}...)
+func (r MetricsRecorder) Impressions(ctx context.Context, obj MetricsObject) {
+	r.impressions.Add(ctx, 1, obj.GetImpressionAttributes()...)
+}
+
+func (r MetricsRecorder) Reasons(ctx context.Context, obj MetricsObject) {
+	r.reasons.Add(ctx, 1, obj.GetReasonAttributes()...)
+}
+
+func (r MetricsRecorder) AddReasonsAndImpressions(ctx context.Context, obj MetricsObject) {
```
Comment (bacherfl) [inline core/pkg/otel/metrics.go:64]: should we rename this to something more generic like e.g. `RecordEvaluation`? Otherwise, if we decide to add further metrics in the future, this method would have to be renamed again as well

File: core/pkg/eval/ievaluator.go (hunk 4 above)
Comment (odubajDT) [reply core/pkg/eval/ievaluator.go]: removed

File: core/pkg/eval/ievaluator.go (hunk 3 above)
Comment (odubajDT) [reply core/pkg/eval/ievaluator.go]: resolved

File: core/pkg/eval/ievaluator.go [hunk 6]
```diff
@@ -2,25 +2,71 @@ package eval
...
+
+func (a AnyValue) HasError() bool {
+	return a.Error != nil
+}
+
+func FeatureFlagReason(val string) attribute.KeyValue {
+	return FeatureFlagReasonKey.String(val)
+}
+
+func ExceptionType(val string) attribute.KeyValue {
```
[the remark this answers is not in the document]
Comment (odubajDT) [reply core/pkg/eval/ievaluator.go]: fixed

**Review by beeme1mr — APPROVED** [review]
This works great from a functionality perspective. I'll let others comment on the code itself.

File: core/pkg/otel/metrics_test.go [hunk 7]
```diff
@@ -90,65 +90,84 @@ func TestNewOTelRecorder(t *testing.T) {
...
+			name: "Reasons",
+			metricFunc: func(exp metric.Reader) {
+				rec := NewOTelRecorder(exp, svcName)
 				for i := 0; i < n; i++ {
-					rec.Impressions(ctx, "key", "variant")
+					rec.Reasons(context.TODO(), "reason")
 				}
 			},
+			metricsLen: 1,
 		},
```
Comment (thisthat) [inline core/pkg/otel/metrics_test.go:154]: I would also add 2 entries for `RecordEvaluation`: one with and one without error

File: core/pkg/otel/metrics.go [hunk 8]
```diff
@@ -100,10 +126,15 @@ func NewOTelRecorder(exporter metric.Reader, serviceName string) *MetricsRecorde
 		"impressions",
 		instrument.WithDescription("The number of evaluation for a given flag"),
 	)
+	reasons, _ := meter.Int64Counter(
+		"reasons",
+		instrument.WithDescription("The number of evaluation for a given reason"),
```
Comment (toddbaert) [inline core/pkg/otel/metrics.go:131]:
```suggestion
		instrument.WithDescription("The number of evaluations for a given reason"),
```

File: core/pkg/otel/metrics.go [hunk 9]
```diff
@@ -46,14 +53,29 @@ func (r MetricsRecorder) InFlightRequestEnd(ctx context.Context, attrs []attribu
...
 		semconv.FeatureFlagVariant(variant),
-		semconv.FeatureFlagProviderName("flagd"),
+		FeatureFlagReason(reason),
+	}...)
+}
+
+func (r MetricsRecorder) Reasons(ctx context.Context, reason string) {
+	r.reasons.Add(ctx, 1, []attribute.KeyValue{
+		semconv.FeatureFlagProviderName(FlagdProviderName),
+		FeatureFlagReason(reason),
```
Comment (toddbaert) [inline core/pkg/otel/metrics.go:68]: In cases where the `Reason` is `"error", we would also have an error code we could add here. It will just be nil or empty in non-error-reason cases. What do you think @thisthat @beeme1mr ?

File: core/pkg/eval/json_evaluator.go [hunk 10]
```diff
@@ -252,7 +251,7 @@ func (je *JSONEvaluator) evaluateVariant(
 		err = jsonlogic.Apply(bytes.NewReader(targetingBytes), bytes.NewReader(b), &result)
 		if err != nil {
 			je.Logger.ErrorWithID(reqID, fmt.Sprintf("error applying rules: %s", err))
-			return "", model.ErrorReason, err
+			return "", model.ErrorReason, errors.New(model.FlagDisabledErrorCode)
```
Comment (Kavindu-Dodan) [inline core/pkg/eval/json_evaluator.go:254]: This is a json logic applying error

```suggestion
			return "", model.ErrorReason, errors.New(model.ParseErrorCode)
```

File: core/pkg/service/flag-evaluation/flag_evaluator.go (hunk 2 above)
Comment (toddbaert) [reply core/pkg/service/flag-evaluation/flag_evaluator.go:287]: Nice catch!

File: core/pkg/eval/json_evaluator.go (hunk 10 above)
Comment (toddbaert) [reply core/pkg/eval/json_evaluator.go:254]: Agreed with @Kavindu-Dodan on this one.

File: core/pkg/otel/metrics.go (hunk 9 above)
Comment (odubajDT) [reply core/pkg/otel/metrics.go:68]: surely, we can add another dimension describing error type, when the reason is error. But will wait if rest of the team agree

Comment (beeme1mr) [reply core/pkg/otel/metrics.go:68]: I think that error codes would be valuable. 👍 

@thisthat any concerns about this?

Comment (odubajDT) [reply core/pkg/otel/metrics.go:70]: already adapted, if it will be declined, I will just revert the last commit

**Review by Kavindu-Dodan — APPROVED** [review]

**Review by thisthat — APPROVED** [review]
